doc/setup-binderhub.rst

```diff
@@ -112,40 +113,74 @@ If you are using ``gcr.io``
 your ``config.yaml`` file::
 
-  registry:
-    prefix:  gcr.io/<google-project-id>/<prefix>
-    enabled: true
+  config:
+    BinderHub:
+      use_registry: true
+      docker_image_prefix: gcr.io/<google-project-id>/<prefix>-
+
 
 .. note::
 
    * **``<google-project-id>``** can be found in the JSON file that you
      pasted above. It is the text that is in the ``project_id`` field. This is
      the project *ID*, which may be different from the project *name*.
    * **``<prefix>``** can be any string, and will be prepended to image names. We
-     recommend something descriptive such as ``binder-dev`` or ``binder-prod``.
+     recommend something descriptive such as ``binder-dev-`` or ``binder-prod-`` (ending with a `-` is useful).
 
 If you are using Docker Hub
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Using Docker Hub is slightly more involved as the registry is not being run
-by the same platform that runs BinderHub.
-
 Update ``config.yaml`` by entering the following::
 
-  registry:
-    enabled: true
-    prefix: <docker-id/organization-name>/<prefix>
-    host: https://registry.hub.docker.com
-    authHost: https://index.docker.io/v1
-    authTokenUrl: https://auth.docker.io/token?service=registry.docker.io
+  config:
+    BinderHub:
+      use_registry: true
+      docker_image_prefix: <docker-id|organization-name>/<prefix>-
 
 .. note::
 
-   * **``<docker-id/organization-name>``** is where you want to store Docker images.
+   * **``<docker-id|organization-name>``** is where you want to store Docker images.
      This can be your Docker ID account or an organization that your account belongs to.
    * **``<prefix>``** can be any string, and will be prepended to image names. We
      recommend something descriptive such as ``binder-dev`` or ``binder-prod``.
 
+If you are using a custom registry
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Authenticating with a Docker registry is slightly more complicated.
+BinderHub knows how to talk to gcr.io and DockerHub,
+but if you are using another registry, you will have to provide more information, in the form of three different urls:
+
+- auth host (added to ``docker/config.json`` from ``registry.host``)
+- registry host (if different from auth host)
+- token url
+
+First, setup the docker configuration with the host used for authentication::
+
+    registry:
+      host: "https://myregistry.io"
+      username: xxx
+      password: yyy
+
+Second, you will need to instruct BinderHub about two additional URLs::
+
+    config:
+      BinderHub:
+        use_registry: true
+        docker_image_prefix: "your-registry.io/<prefix>-"
+      DockerRegistry:
+        token_url: "https://myregistry.io/v2/token?service="
+        registry_host: "https://registry.myregistry.io"
+
+The two URLs will come from your registry.
+``registry_host`` only needs to be specified if it is different
+from the ``registry.host`` that you specified above.
+
+The ``registry.host`` is the host used in docker ``config.json``,
```

Comment:
Make this paragraph a note. `...can differ when the auth host and registry host differ.`

Reply:
Ended up reworking this entirely, since the extra URL should be a super special case, I think.